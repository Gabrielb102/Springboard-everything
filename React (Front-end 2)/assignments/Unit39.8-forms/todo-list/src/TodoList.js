import {useState} from "react";
import { v4 as uuid } from 'uuid';
import Todo from "./Todo";
import TodoForm from "./TodoForm";

const TodoList = () => {
    const [ todos, setTodos ] = useState([]);

    // BUG: in the removeFunc function, todos is shortened to include everything BEFORE THE clicked component
    // The state is not being read correctly, and I cannot get it
    const removeFunc = (id) => {
        debugger;
        console.log('todos: ', todos);
        const todoToDie = todos.find((todo) => {
            const match = todo.props.id === id;
            return match;
        });

        // if (todoToDie) {
        //     const idx = todos.indexOf(todoToDie);
        //     setTodos(todos => {
        //         todos = todos.splice(idx, 1);
        //         return todos;
        //     });

        // } else {
        //     return todos;
        // }
    }

    const addTodo = (task) => {
        const id = uuid();
        const newTodo = <Todo key={id} id={id} removeFunc={() => removeFunc(id)} task={task} />;
        setTodos([newTodo, ...todos]);
    }

    console.log(todos)
    return (
        <>  
            <h1>Your To Do List for To Day</h1>
            <TodoForm addTodo={addTodo} />
            {todos.map(todo => <li>{todo}</li>)}
        </>
    )
}

export default TodoList;