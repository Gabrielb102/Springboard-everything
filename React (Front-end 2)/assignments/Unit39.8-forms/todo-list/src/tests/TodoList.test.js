import TodoList from "../TodoList";
import {render, fireEvent, queryByText} from "@testing-library/react";

describe("Tests for the TodoList component", () => {

    test("Renders", () => {
        render(<TodoList/>);
    })

    test("Renders the same UI consistently", () => {
        const { asFragment } = render(<TodoList />);
        expect(asFragment).toMatchSnapshot();
    })

    test("Adds todo item when form is submitted", () => {
        //Generate todo
        const { getByPlaceholderText, queryByText } = render(<TodoList />);
        const input = getByPlaceholderText("New Task");
        const button = queryByText("Post")
        fireEvent.change(input, {target: {value: "Groceries"}});
        fireEvent.click(button);

        //Check for todo
        const todo = queryByText("Groceries");
        expect(todo).not.toBeNull()
    })

    test("Clicking x removes specified todo list item", () => {
        const { getByPlaceholderText, queryAllByText } = render(<TodoList />);
        const input = getByPlaceholderText("New Task");
        const button = queryByText("Post")

        //Generate three todos
        fireEvent.change(input, {target: {value: "Groceries"}});
        fireEvent.click(button);
        fireEvent.change(input, {target: {value: "Walk Waldo"}});
        fireEvent.click(button);
        fireEvent.change(input, {target: {value: "Eat Boody"}});
        fireEvent.click(button);

        //Click x
        const xs = queryAllByText("x");
        fireEvent.click(xs[1]);

        //check for todo
        const todo1 = queryByText("Groceries");
        expect(todo1).not.toBeNull();
        const todo2 = queryByText("Walk Waldo");
        expect(todo2).toBeNull();
        const todo3 = queryByText("Eat Boody");
        expect(todo3).not.toBeNull();
    })
})