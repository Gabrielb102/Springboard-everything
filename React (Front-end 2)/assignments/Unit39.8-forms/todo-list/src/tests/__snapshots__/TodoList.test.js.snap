// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests for the TodoList component Renders the same UI consistently 1`] = `[Function]`;
